Tests area's optimization.ini generation

guilpier-code opened the pull request.

guilpier-code committed:
Changing write of ini file : taking care of areas optimization.ini

guilpier-code committed:
Changing write of ini file : taking care of properties.ini (for links)

guilpier-code committed:
Improving optimization.ini generation : correction due to sonarQube analysis

guilpier-code committed:
Improving properties.ini generation : correction due to sonarQube analysis

guilpier-code committed:
Test links property save : first commit (do not compile)

guilpier-code committed:
Test links property save : testing default link

guilpier-code committed:
Test links property save : refactoring : introducing a class output folder finder

This output folder finder answers question :  was the output folder found ?
This question is then tested with a boost check

guilpier-code committed:
[skip ci] Test links property save : some cleaning

guilpier-code committed:
Test links property save : change strategy --> tests contain the reference file content

guilpier-code committed:
Test links property save : add 2 kind of tests

- tests where generated properties.ini is compared to a reference properties.ini (of which content is set in test)
- tests where we check that the generated properties.ini contains an expected line associated to a particular property.

guilpier-code committed:
Improving optimization.ini generation : extracting code that generates an area's "optimization.ini" into a function

guilpier-code committed:
[skip ci] Test links property save : clean the cmake file

guilpier-code committed:
Test area's optimization.ini generation : add 2 tests
- Testing content of optimization.ini with area's default values
- Testing content of optimization.ini with none default values

guilpier-code committed:
[skip ci] Test area's optimization.ini generation : improvement on generated and reference .ini files names

guilpier-code committed:
Test links property save :  close ini files when opened, in any case

guilpier-code committed:
Test area's optimization.ini generation : adding 6 more tests
- 4 tests on nodal optimization parameter
- 1 test on unsupplied energy cost
- 1 test on spilled energy cost

guilpier-code committed:
[skip ci] Test links property save : avoid code duplication in ini files names

guilpier-code committed:
Test area's optimization.ini generation : 2 more tests + functions changed
- adding 2 tests on filters (synthesis and year by year)
- function that compares files and function that finds a string in a file were implemented in another way

guilpier-code committed:
Test links property save : functions implemented in another way

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/test-links-property-save' into feature/tests-area-ini-optimization-dump

# Conflicts:
#	src/tests/src/libs/antares/study/area/CMakeLists.txt

guilpier-code committed:
Tests on ini files : factoring code about files handling (removal, comparison, content)

flomnes commented on src/tests/src/libs/antares/study/area/files-helper.h:
```diff
@@ -0,0 +1,50 @@
+	string content_2 = readFileIntoString(fileName_2);
+	if (content_1.compare(content_2))
+		return false;
+	return true;
```
Consider using `operator==` would make this code much more readable.

`return content_1 == content_2;`

guilpier-code replied:
ok for me, done

guilpier-code commented on src/tests/src/libs/antares/study/area/files-helper.h:
```diff
@@ -0,0 +1,50 @@
+#pragma once
```
Currently, this file is not inside the IDE solution.
It should be (if not, a search over the entire solution would avoid searching in this file)

flomnes commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -339,6 +305,26 @@ static bool AreaListSaveToFolderSingleArea(const Area& area, Clob& buffer, const
+
+    if (!ini.save(buffer))
+        return false;
+    return true;
```
`return ini.save(buffer);`

guilpier-code replied:
I agree, done

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -635,86 +636,32 @@ bool AreaLinksSaveToFolder(const Area* area, const char* const folder)
-    }
-    file << buffer;
 
     return ret;
```
`return ini.save(filename);`

guilpier-code replied:
on this, I don't agree : look at **ret** variable, and how it is used and returned

flomnes replied:
ok

flomnes commented on src/libs/antares/study/links/property-links-helper.h:
```diff
@@ -0,0 +1,74 @@
+            to_return = "other";
+            break;
+        }
+        return to_return;
```
```cpp
case Dat::atAC : return "ac";
case Data::atDC : return "dc";
// etc.
```

`break` not required since `return` stops the function's execution.

guilpier-code replied:
ok, done

flomnes commented on src/libs/antares/study/links/property-links-helper.h:
```diff
@@ -0,0 +1,74 @@
+        {
+        case Data::stDot:
+            to_return = "dot";
+            break;
```
Same remark here

guilpier-code replied:
done

flomnes commented on src/tests/src/libs/antares/study/area/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+							libmodel_antares
+)
+
+# Storing test-save-link-properties under the folder Unit-tests in the IDE
+set_target_properties(test-save-link-properties PROPERTIES FOLDER Unit-tests)
```
I'm not sure folder **Unit-tests** exists. Is it generated by some IDEs ?

guilpier-code replied:
At least, this solution "subfolder" is generated on MSVC

flomnes commented on src/tests/src/libs/antares/study/area/files-helper.h:
```diff
@@ -0,0 +1,50 @@
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
```
Good practice : avoid defining helper namespaces in header files. Create a separate .cpp for function implementations and define them here.

flomnes commented on src/tests/src/libs/antares/study/area/files-helper.h:
```diff
@@ -0,0 +1,50 @@
+	{
+		fs::path fileToRemove = filesToRemove[i];
+		if (exists(fileToRemove))
+			remove(fileToRemove);
```
`fs::exists` and `fs::remove`

guilpier-code replied:
done

flomnes commented on src/tests/src/libs/antares/study/area/test-save-area-optimization-ini.cpp:
```diff
@@ -0,0 +1,324 @@
+	Area* area = study->areaAdd("Area");
+
+	string generatedIniFileName = "optimization.ini";
+	string referenceIniFileName = "optimization-reference.ini";
```
Since `generatedIniFileName` and `generatedIniFileName` are const and always have the same value, consider using global variables defined at the beginning of this source file.

guilpier-code replied:
done

flomnes commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -339,6 +305,26 @@ static bool AreaListSaveToFolderSingleArea(const Area& area, Clob& buffer, const
+    IniFile ini;
+    IniFile::Section* section = ini.addSection("nodal optimization");
+
+    section->add("non-dispatchable-power", (bool)(area.nodalOptimization & anoNonDispatchPower));
```
Consider using `static_cast<bool>(...)` instead. See [this SO answer](https://stackoverflow.com/a/18414126) for more details.

guilpier-code replied:
you're right

flomnes commented on src/libs/antares/study/filter.cpp:
```diff
@@ -33,6 +33,44 @@ namespace Antares
+
+std::string filterIntoString(uint filter)
+{
+    std::string to_return;
```
Choose more expressive function & variable names. For example:

`filterIntoString` -> `datePrecisionIntoString`

`to_return` -> `datePrecision`

This characteristic is referred to as "date precision" or "precision level" in the rest of the code.

guilpier-code replied:
done

flomnes commented on src/libs/antares/study/links/property-links-helper.h:
```diff
@@ -0,0 +1,74 @@
+            break;
+		}
+		return to_return;
+	}
```
```cpp
case Data::tncEnabled : return "enabled";
// etc.
```

guilpier-code replied:
ok

guilpier-code committed:
[skip ci] Tests in ini files : corrections resulting from a code review (first part)

guilpier-code committed:
Tests in ini files : corrections resulting from a code review (part 2)

guilpier-code committed:
Tests in ini files : corrections resulting from a code review (part 3)

guilpier-code committed:
Tests in ini files : corrections due to a CI failure (Linux)

guilpier-code committed:
Tests on properties.ini file : corrections due to sonarQube report
- no "using namespace" in header file
- avoid code duplication as much as possible

guilpier-code committed:
Tests on properties.ini file : corrections due to more sonarQube issues

guilpier-code committed:
Tests on optimization.ini file : removing some code duplication
We do this by :
- Introducing a fixture (setup + tear down)
- Introducing a class for the reference ini file (to be compared with the generated file)

guilpier-code committed:
Tests on ini file : correction due to sonarrQube analysis
